### PR TITLE
Add init v4 plugins based on devfile v2

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
This PR add init v4 for plugins based on devfile v2, it's just Che Theia and machine exec extracted from https://github.com/devfile/devworkspace-operator/blob/master/samples/flattened_theia-next.yaml

Tested API:
<details>
<summary>https://plugin-registry-eclipse-che.192.168.39.241.nip.io/v4/plugins/index.json</summary>

```json
[
  {
    "id": "eclipse/che-theia/next",
    "description": "Eclipse Theia",
    "displayName": "theia-ide",
    "links": {
      "self": "/v4/plugins/eclipse/che-theia/next"
    },
    "name": "che-theia",
    "publisher": "eclipse",
    "type": "Che Editor",
    "version": "latest"
  },
  {
    "id": "eclipse/che-machine-exe/nightly",
    "description": "Che Plug-in with che-machine-exec service to provide creation terminal or tasks for Eclipse Che workspace containers.",
    "displayName": "Che machine-exec Service",
    "links": {
      "self": "/v4/plugins/eclipse/che-machine-exec/nightly"
    },
    "name": "che-machine-exec-plugin",
    "publisher": "eclipse",
    "type": "Che Plugin",
    "version": "nightly"
  }
]
```

</details>

<details>
<summary>https://plugin-registry-eclipse-che.192.168.39.241.nip.io/v4/plugins/eclipse/che-machine-exec/nightly/</summary>

```yaml
schemaVersion: "2.0.0"
metadata:
  name: "Che Machine Exec"
components:
- name: terminal
  attributes:
    "app.kubernetes.io/name": che-terminal.eclipse.org
    "app.kubernetes.io/part-of": che.eclipse.org
    "app.kubernetes.io/component": terminal
  container:
    image: "quay.io/eclipse/che-machine-exec:nightly"
    command: ['/go/bin/che-machine-exec']
    args:
      - '--url'
      - '0.0.0.0:4444'
      - '--pod-selector'
      - controller.devfile.io/workspace_id=$(DEVWORKSPACE_ID)
    endpoints:
      - name: "che-mach-exec"
        exposure: public
        targetPort: 4444
        protocol: ws
        secure: false
        attributes:
          type: terminal
```

</details>

### What issues does this PR fix or reference?
It's done as a first step of https://github.com/eclipse/che/issues/19034

### How to test this PR?
You use `quay.io/sleshche/che-plugin-registry:v4` to test v4.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: v4/plugins/eclipse/che-machine-exec/nightly/devfile.yaml
Comment (sleshchenko) [inline v4/plugins/eclipse/che-machine-exec/nightly/devfile.yaml:7]: I'm not sure about its usage and who is supposed to provision that attributes, but it's not big deal. It's here at least for time being as David and Mario provided

File: v4/plugins/eclipse/che-machine-exec/nightly/devfile.yaml
Comment (sleshchenko) [inline v4/plugins/eclipse/che-machine-exec/nightly/devfile.yaml:23]: Note that it's marked as unsecure by purpose, since it's not implemented yet on Che side for DevWorkspace

And it listen to 0.0.0.0 which is again different from v3 due the same reason, DevWorkspace does not have proxy that would expose localhost to 0.0.0.0 in a secure manner.

**Review by benoitf — CHANGES_REQUESTED** [review]
there shouldn't be any v4 folder added by hands

input comes from 3 files stored at the root folder

File: v4/plugins/eclipse/che-theia/next/devfile.yaml
Comment (benoitf) [inline v4/plugins/eclipse/che-theia/next/devfile.yaml:1]: it should be generated from che-editors.yaml file

File: v4/plugins/eclipse/che-machine-exec/nightly/devfile.yaml
Comment (benoitf) [inline v4/plugins/eclipse/che-machine-exec/nightly/devfile.yaml:1]: should be generated from che-plugins.yaml file

File: build/dockerfiles/Dockerfile
Comment (benoitf) [inline build/dockerfiles/Dockerfile:28]: we don't need v4 if one are meta.yaml and the other one are devfile.yaml

Comment (sleshchenko) [reply build/dockerfiles/Dockerfile:28]: I think we need it.
There is no technical need, but it seems reasonable to me, to make it clear that the API is changed.

**Comment by sleshchenko** [comment]
> there shouldn't be any v4 folder added by hands
> input comes from 3 files stored at the root folder

Fully agree. And I expected plugin team to fix the described issue, it's only my initial contribution made in a hour to unblock Dashboard from moving forward in DevWorkspace direction.
Feel free to close the PR if you think we must not merge it even as the first step.

**Review by amisevsk — COMMENTED** [review]
We should have a forked development plugin registry for use until devfile 2.0 plugin generation is implemented. I'll work on setting something up, it shouldn't be too complicated.

**Comment by ericwill** [comment]
> We should have a forked development plugin registry for use until devfile 2.0 plugin generation is implemented. I'll work on setting something up, it shouldn't be too complicated.

+1 to that

**Comment by sleshchenko** [comment]
That's not going to be merged, so closing it.
Until the proper solution is not implemented, Dashboard will depends on the custom plugin registry.
`chectl server:deploy --workspace-engine=devworkspace` will depends on custom plugin registry as well.

**Comment by benoitf** [comment]
> We should have a forked development plugin registry

@amisevsk it could be just a github repository with these 3 files but let's communicate on the files you're using and changes that you've made so the generator could handle that.

Also probably need to patch https://github.com/eclipse/che-machine-exec/blob/6dbaaa9037242dc0409832e15a159ddf24653b2d/cfg/cfg.go#L92-L96 to avoid to give extra parameter.

'proper' solution could be achieved before milestone 1 as well if machine-exec is patched to handle DEVWORKSPACE_ID as well.

**Comment by sleshchenko** [comment]
>  it could be just a github repository with these 3 files

I don't see how it's possible if we want `chectl server:deploy --workspace-engine=devworkspace` where Che supports Che workspace and DevWorkspace at the same time.

**Comment by sleshchenko** [comment]
> machine-exec is patched to handle DEVWORKSPACE_ID as well.

The draft PR is already in place https://github.com/eclipse/che-machine-exec/pull/130

**Comment by benoitf** [comment]
I've created https://github.com/eclipse/che/issues/19060

**Pull request closed without merging**
